vllm/core/block_manager_v2.py

```diff
@@ -113,27 +121,47 @@ def can_allocate(self, seq_group: SequenceGroup) -> AllocStatus:
         else:
             return AllocStatus.LATER
 
+    def _allocate_sequence(self, seq: Sequence) -> BlockTable:
+        block_table = BlockTable(
+            block_size=self.block_size,
+            block_allocator=self.block_allocator,
+        )
+        assert self.block_sliding_window is None
+        block_table.allocate(seq.get_token_ids())
+
+        return block_table
+
     def allocate(self, seq_group: SequenceGroup) -> None:
+        # Allocate self-attention block tables for decoder sequences
         waiting_seqs = seq_group.get_seqs(status=SequenceStatus.WAITING)
         assert not (set(seq.seq_id for seq in waiting_seqs)
                     & self.block_tables.keys()), "block table already exists"
 
         # NOTE: Here we assume that all sequences in the group have the same
         # prompt.
         seq = waiting_seqs[0]
-
-        block_table = BlockTable(
-            block_size=self.block_size,
-            block_allocator=self.block_allocator,
-        )
-        assert self.block_sliding_window is None
-        block_table.allocate(seq.get_token_ids())
+        block_table: BlockTable = self._allocate_sequence(seq)
         self.block_tables[seq.seq_id] = block_table
 
         # Assign the block table for each sequence.
         for seq in waiting_seqs[1:]:
             self.block_tables[seq.seq_id] = block_table.fork()
 
+        # Allocate cross-attention block table for encoder sequence
+        #
+        # NOTE: Here we assume that all sequences in the group have the same
+        # encoder prompt.
+        request_id = seq_group.request_id
+
+        assert (request_id
+                not in self.cross_block_tables), \
+                "block table already exists"
+
+        encoder_seq = seq_group.get_encoder_seq()
+        if encoder_seq is not None:
```

Comment:
Do we want to add a check here somewhere that prefix caching is disabled for encoder seq. In L:81 where the block_allocator is initialized it could be initialized in the prefix_caching mode?

Reply:
Agreed that at initialization time block_allocator could be constructed with prefix_caching enabled in principle.

Currently the Block Manager v2 constructor (where block allocator is constructed) does not know whether the model is encoder/decoder or not; under the current implementation, this information is not revealed until a SequenceGroup arrives with a non-None encoder_seq.

Therefore the location you highlighted, within the allocate() method, is a proper place to check whether an encoder_seq is present in the SequenceGroup argument, and then use this to raise NotImplemented exceptions if prefix caching or sliding window attention (SWA) are enabled. I have introduced this change.

Per previous suggestion, I have modified allocate() so that "not decoder_only" rather than "encoder_seq is not None" is the method used to check for the encoder/decoder scenario.